### PR TITLE
Esa gaia gaiapcr 1290 include the new data link linking parameter

### PR DESCRIPTION
**Pull request by cosmoJFH**
Hi, my name is Jorge Fdez, and I work as a software developer at ESAC (gaia mission). The new parameter linking_parameter can be used in the datalink service, in order to provide an additional meaning to the source identifiers: source_id, transit_id and image_id. This parameter is optional, in order to be backward compatible, and therefore, if the parameter is not set, the source identifiers are considered as source_id,

These changes are related to the jira ticket https://s2e2.cosmos.esa.int/jira/browse/GAIAPCR-1290. 

@esdc-esac-esa-int

### CODE REVIEW

**Comment by cosmoJFH** [comment]
There exist a failing test that is not part of thisPR

image

File: astroquery/gaia/core.py [hunk 1]
```diff
@@ -228,22 +235,25 @@ def load_data(self, ids, *, data_release=None, data_structure='INDIVIDUAL',
 
         output_file_specified = False
         if output_file is None:
-            output_file = os.path.join(os.getcwd(), temp_dirname, downloadname_formated)
+            output_file = os.path.join(os.getcwd(), temp_dirname,
+                                       downloadname_formated)
```
Comment (bsipocz) [inline astroquery/gaia/core.py:239]: We do let longer lines in astroquery (`max-line-length = 120`). It is supposed to be picked up from the config files, but I'm happy to add new sections that your editor/IDE would respect.

File: astroquery/gaia/core.py [hunk 2]
```diff
@@ -194,6 +196,11 @@ def load_data(self, ids, *, data_release=None, data_structure='INDIVIDUAL',
             For GAIA DR4, possible values will be ['EPOCH_PHOTOMETRY', 'RVS', 'XP_CONTINUOUS', 'XP_SAMPLED',
             'MCMC_GSPPHOT', 'MCMC_MSC', 'EPOCH_ASTROMETRY', 'RV_EPOCH_SINGLE', 'RV_EPOCH_DOUBLE', 'RVS_EPOCH' or
             'RVS_TRANSIT']
+        linking_parameter : str, optional, default None, valid values: SOURCE_ID, TRANSIT_ID, IMAGE_ID
+            By default, all the identifiers are considered as source_id
+            SOURCE_ID: the identifiers are considered as source_id
```
Comment (bsipocz) [inline astroquery/gaia/core.py:201]: Why to have `SOURCE_ID` as the default then, instead of None?

**Review by bsipocz — COMMENTED** [review]
Thank you for the PR @cosmoJFH.

I left a couple of inline comments. Besides those, it would be great not to include any reformatting changes, like the line wraps in unrelated PRs. Now it's just adds a couple 10s of extra diff, but in bigger PRs it can be rather distruptive.

Do get this merged, would you consider showing the usage in the narrative docs? Either by adding a new example, or modifying one of the existing ones.

Also, having a non monkeypatched test would be useful, so we actually test that the code works with the server, as expected.

File: astroquery/gaia/core.py (hunk 1 above)
Comment (cosmoJFH) [reply astroquery/gaia/core.py]: Sorry for the mistake, I misconfigured my pycharm IDE: I set the  max line length to 79 characters. I have reverted the formatting changes

File: astroquery/gaia/core.py (hunk 2 above)
Comment (cosmoJFH) [reply astroquery/gaia/core.py:199]: Yes, you are right it is more convenient to set the default value to  SOURCE_ID

**Comment by cosmoJFH** [comment]
We have included a new section (2.7) to the gaia documentation, devoted to the datalink service. Note that it does not include any reference to the new parameter _linking_parameter_, since this information must be included in the gaia help (https://www.cosmos.esa.int/web/gaia-users/archive0 beforehand by the gaia team as ESAC.

File: docs/gaia/gaia.rst [hunk 3]
```diff
@@ -851,6 +852,45 @@ will be able to access to your shared table in a query.
...
+The DataLink products are publicly accessible via Astroquery.Gaia using the *load_data* method. The following example shows how to retrieve the DataLink products associated to three sources in Gaia DR3:
+
+
+.. code-block:: python
+
+  >>> retrieval_type = 'ALL'          # Options are: 'EPOCH_PHOTOMETRY', 'MCMC_GSPPHOT', 'MCMC_MSC', 'XP_SAMPLED', 'XP_CONTINUOUS', 'RVS', 'ALL'
+  >>> data_structure = 'COMBINED'     # Options are: 'INDIVIDUAL', 'COMBINED', 'RAW'
+  >>> data_release   = 'Gaia DR3'     # Options are: 'Gaia DR3' (default), 'Gaia DR2'
+  >>> datalink = Gaia.load_data(ids=[2263166706630078848, 2263178457660566784, 2268372099615724288], data_release = data_release, retrieval_type=retrieval_type, data_structure = data_structure, verbose = False, output_file = None)
+
```
Comment (bsipocz) [inline docs/gaia/gaia.rst:875]: These will need to be tested, and a little bit reformatted.
```suggestion
.. doctest-remote-data::

  >>> retrieval_type = 'ALL'          # Options are: 'EPOCH_PHOTOMETRY', 'MCMC_GSPPHOT', 'MCMC_MSC', 'XP_SAMPLED', 'XP_CONTINUOUS', 'RVS', 'ALL'
  >>> data_structure = 'COMBINED'     # Options are: 'INDIVIDUAL', 'COMBINED', 'RAW'
  >>> data_release   = 'Gaia DR3'     # Options are: 'Gaia DR3' (default), 'Gaia DR2'
  >>> datalink = Gaia.load_data(ids=[2263166706630078848, 2263178457660566784, 2268372099615724288], 
  ...                           data_release=data_release, retrieval_type=retrieval_type, data_structure=data_structure)
  >>> datalink.keys()
  dict_keys(['MCMC_GSPPHOT_COMBINED.xml', 'EPOCH_PHOTOMETRY_COMBINED.xml', 'RVS_COMBINED.xml', 'MCMC_MSC_COMBINED.xml', 'XP_SAMPLED_COMBINED.xml', 'XP_CONTINUOUS_COMBINED.xml'])

```

File: docs/gaia/gaia.rst [hunk 4]
```diff
@@ -851,6 +852,45 @@ will be able to access to your shared table in a query.
...
+The DataLink products are stored inside a Python Dictionary. Each of its elements (keys) contains a one-element list that can be extracted as follows:
+.. code-block:: python
+
+  >>> dl_keys  = [inp for inp in datalink.keys()]
+  >>> dl_keys.sort()
+  >>> print(f'The following Datalink products have been downloaded:')
+  >>> for dl_key in dl_keys:
+        print(f' * {dl_key}')
```
Comment (bsipocz) [inline docs/gaia/gaia.rst:883]: I would say to drop all of this, the example above should be sufficient enough for python users.
```suggestion

```

File: docs/gaia/gaia.rst [hunk 5]
```diff
@@ -851,6 +852,45 @@ will be able to access to your shared table in a query.
...
+3. Datalink service (Public and Authenticated)
+---------------------------
+
+
+DataLink_ is a data access protocol compliant with the IVOA_ architecture that provides a linking mechanism between
+datasets offered by different services. In practice, it can be seen and used as a web service providing the list of additional
+data products available for each object outside the main catalogue(s). For more information about the products served via
+DataLink in the Gaia ESA Archive we recommend to read the Archive DataLink tutorials available at https://www.cosmos.esa.int/web/gaia-users/archive/datalink-products.
+
+The DataLink products are publicly accessible via Astroquery.Gaia using the *load_data* method. The following example shows how to retrieve the DataLink products associated to three sources in Gaia DR3:
```
Comment (bsipocz) [inline docs/gaia/gaia.rst:866]: Do link to the API docs whenever possible. Also, moved the mention of the dictionary result from the dropped example below.

```suggestion
The DataLink products are publicly accessible via the `~astroquery.gaia.Gaia.load_data` method. 
The following example shows how to retrieve the DataLink products associated with three sources in Gaia DR3 as a python dictionary:
```

**Review by bsipocz — COMMENTED** [review]
Some suggestions of wordsmithing and cleanups in the docs, otherwise this is good to go. 

@cosmoJFH - if you're happy with the suggestions, just give me the go-ahead, and I can commit the changes along with fixes to the docs build and merge.

File: docs/gaia/gaia.rst (hunk 3 above)
Comment (cosmoJFH) [reply docs/gaia/gaia.rst:875]: We are happy with your changes.

File: docs/gaia/gaia.rst (hunk 4 above)
Comment (cosmoJFH) [reply docs/gaia/gaia.rst:883]: We are happy with your changes.

File: docs/gaia/gaia.rst (hunk 5 above)
Comment (cosmoJFH) [reply docs/gaia/gaia.rst:866]: We are happy with your changes.

**Comment by cosmoJFH** [comment]
Hi @bsipocz we are happy with the changes you suggested for the documentation.

**Review by bsipocz — APPROVED** [review]

**Comment by bsipocz** [comment]
Thank you @cosmoJFH!

**Pull request merged**